File: Sources/HaviKit/HaviRedaction.swift
import Foundation

/// Redaction policy (design §1). Privacy-first / default-mask: every SwiftUI
/// `TextField` / `SecureField` is masked in the frozen snapshot unless
/// explicitly `.haviReveal()`-ed. The reading passage, OCR previews, and any
/// child-name view are the integration's responsibility to tag `.haviRedacted()`
/// (design §7) — the default `TextField` mask is a safety net, not a guarantee.
public struct HaviRedactionPolicy: Sendable {
    public var maskTextFieldsByDefault: Bool

    public init(maskTextFieldsByDefault: Bool = true) {
        self.maskTextFieldsByDefault = maskTextFieldsByDefault
    }
}

/// The KEY-name secret scrub, ported verbatim from
/// `assets/shared/enrichment.js` (`scrubSecrets` / `isSecretKey`). It is a
/// **key-name denylist** (matching the repo), not a value-pattern scanner:
/// every object key whose name matches the regex is dropped at any nesting
/// depth. Applied on device to `x:havi.contexts` and `x:havi.tags` before send
/// (design §3) so secrets captured via `setContext` / `setTag` never reach the
/// stored annotation.
public enum HaviRedaction {
    /// `token|secret|password|api[_-]?key|authorization|cookie`, case-insensitive
    /// — byte-identical to `SECRET_KEY_RE` in `enrichment.js`.
    static let secretKeyPattern = "token|secret|password|api[_-]?key|authorization|cookie"

    public static func isSecretKey(_ key: String) -> Bool {
        key.range(of: secretKeyPattern, options: [.regularExpression, .caseInsensitive]) != nil
    }

    /// Deep, non-mutating copy with every secret-shaped key dropped at any depth.
    /// Objects and arrays recurse; scalars pass through unchanged — mirroring the
    /// JS `scrubSecrets` recursion (objects, arrays, else passthrough).
    public static func scrub(_ value: Any) -> Any {
        if let dictionary = value as? [String: Any] {
            var out: [String: Any] = [:]
            for (key, nested) in dictionary {
                if isSecretKey(key) { continue }
                out[key] = scrub(nested)
            }
            return out
        }
        if let array = value as? [Any] {
            return array.map { scrub($0) }
        }
        return value
    }
}

File: android/havikit-core/src/main/kotlin/ai/handgemacht/havikit/HaviRedaction.kt
package ai.handgemacht.havikit

/**
 * Redaction policy (wire spec §1). Privacy-first / default-mask: every text input
 * is blacked out in the frozen snapshot unless explicitly revealed. The Android
 * capture layer (later stage) consumes [maskTextFieldsByDefault]; Stage 1 carries
 * it so `HaviConfig` matches the iOS shape.
 */
public data class HaviRedactionPolicy(
    val maskTextFieldsByDefault: Boolean = true,
)

/**
 * The KEY-name secret scrub, byte-identical to the web `enrichment.js`
 * `SECRET_KEY_RE` and the iOS `HaviRedaction` (wire spec §6.6). A key-name
 * denylist — not a value scanner: any object key whose name matches the regex
 * (case-insensitive, unanchored substring) is dropped at any nesting depth.
 * Arrays recurse; scalars pass through. Applied to `x:havi.contexts` and
 * `x:havi.tags` before send.
 */
public object HaviRedaction {
    /** `token|secret|password|api[_-]?key|authorization|cookie`, case-insensitive. */
    public const val SECRET_KEY_PATTERN: String = "token|secret|password|api[_-]?key|authorization|cookie"

    private val secretKeyRegex = Regex(SECRET_KEY_PATTERN, RegexOption.IGNORE_CASE)

    public fun isSecretKey(key: String): Boolean = secretKeyRegex.containsMatchIn(key)

    /**
     * Deep, non-mutating copy with every secret-shaped key dropped at any depth.
     * Objects and arrays recurse; scalars pass through unchanged.
     */
    public fun scrub(value: Any?): Any? =
        when (value) {
            is Map<*, *> -> {
                val out = LinkedHashMap<String, Any?>()
                for ((rawKey, nested) in value) {
                    val key = rawKey as? String ?: continue
                    if (isSecretKey(key)) continue
                    out[key] = scrub(nested)
                }
                out
            }

            is List<*> -> value.map { scrub(it) }

            else -> value
        }
}
